File: scala-example/spring-boot-example/src/main/scala/com/github/poad/scala/springboot/example/service/MessageServiceImpl.scala
package com.github.poad.scala.springboot.example.service
import java.util.Optional

import com.github.poad.scala.springboot.example.content.Message
import com.github.poad.scala.springboot.example.entity.MessageEntity
import com.github.poad.scala.springboot.example.repository.MessageRepository
import javax.ws.rs.WebApplicationException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class MessageServiceImpl(@Autowired private val repository: MessageRepository) extends MessageService {
  import scala.collection.JavaConverters._
  override def all(): Seq[Message] = repository.findAll().asScala.
    map(entity => Message(entity.id, entity.message)).toSeq

  override def byId(id: String): Optional[Message] =
    repository.findById(id).map(entity => Message(entity.id, entity.message))

  override def create(message: Message): String = {
    val id = repository.uuid()
    val entity = MessageEntity(message.message)
    entity.id = id
    repository.save(entity)
    id
  }

  override def update(id: String, message: Message): Optional[Message] = {
    val entity = repository.findById(id)
    if (entity.isPresent) {
      repository.save(entity.map(entity => entity.message = message.message).get())
    }
    entity.map(e => Message(id, message.message))
  }

  override def deleteById(id: String): Optional[Message] = {
    val entity = repository.findById(id)
    if (entity.isPresent) {
      repository.delete(entity.get())
    }
    entity.map(entity => Message(entity.id, entity.message))
  }
}

File: java-maven-example/spring-boot-minimal-jdk/src/main/java/com/github/poad/example/service/MessageServiceImpl.java
package com.github.poad.example.service;

import com.github.poad.example.domain.Message;
import com.github.poad.example.entity.MessageEntity;
import com.github.poad.example.repository.MessageRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.annotation.Immutable;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

@Immutable
@Component
public class MessageServiceImpl implements MessageService {

    private final MessageRepository repository;

    public MessageServiceImpl(@Autowired MessageRepository repository) {
        this.repository = repository;
    }

    @Override
    public List<Message> findAll() {
        return repository.findAll().stream()
                .map(entity -> new Message(entity.getId().toString(), entity.getMessage()))
                .collect(Collectors.toList());
    }

    @Override
    public Optional<Message> findById(String id) {
        return repository.findById(UUID.fromString(id))
                .map(entity -> new Message(entity.getId().toString(), entity.getMessage()));
    }

    @Override
    public Message create(String message) {
        MessageEntity entity = repository.saveAndFlush(new MessageEntity(UUID.nameUUIDFromBytes(repository.uuid()), message));
        return new Message(entity.getId().toString(), entity.getMessage());
    }

    @Override
    public Optional<Message> update(String id, String message) {
        return repository.findById(UUID.fromString(id))
                .map(entity -> repository.saveAndFlush(new MessageEntity(UUID.fromString(id), message)))
                .map(entity -> new Message(entity.getId().toString(), entity.getMessage()));
    }

    @Override
    public void delete(String id) {
        repository.findById(UUID.fromString(id)).map(MessageEntity::getId).ifPresent(repository::deleteById);
    }


}
